# Introduction to Mathematical Optimization Modeling

## Objective

The below problem is one of the introductory modeling practice examples offered by [Gurobi - Jupyter Modeling Examples](https://www.gurobi.com/jupyter_models/?_difficulty_level=introductory).

The "Problem Description", "Mathematical Optimization", and the majority of codes and schematics are directly from the source.
However, significant portion of the comments within the main text here are written by myself.

I took an optimization course called "Deterministic Optimization"[(course description link)](https://catalog.gatech.edu/search/?P=ISYE%206669) from Georgia Tech from professor Renato Monteiro as part of MS in Operations Research program [(program description link)](https://www.isye.gatech.edu/academics/masters/degrees/ms-operations-research) in the fall of 2024. I successfully finished the course and received an A. This is to review the concepts that I studied and to learn how to implement them in python codes using gurobipy.

## Problem description

Consider a consulting company that has three open positions: Tester, Java Developer, and Architect. The three top candidates (resources) for the positions are: Carlos, Joe, and Monika. The consulting company administered competency tests to each candidate in order to assess their ability to perform each of the jobs. The results of these tests are called *matching scores*. Assume that only one candidate can be assigned to a job, and at most one job can be assigned to a candidate.

The problem is to determine an assignment of resources and jobs such that each job is fulfilled, each resource is assigned to at most one job, and the total matching scores of the assignments is maximized.

## Mathematical optimization 

Mathematical optimization (which is also known as mathematical programming) is a declarative approach where the modeler formulates an  optimization problem that captures the key features of a complex decision problem. The Gurobi Optimizer solves the mathematical optimization problem using state-of-the-art mathematics and computer science.

A mathematical optimization model has five components:

* Sets
* Parameters
* Decision variables
* Constraints
* Objective function(s)

In [1]:
%pip install gurobipy

Note: you may need to restart the kernel to use updated packages.


The following code imports the Gurobi and the GRB class. Gurobi API is used for creating models, constraints, and objective functions. (e.g., ``gp.Model()``) GRB class is used for setting the types for variables (e.g., `GRB.CONTINUOUS`, `GRB.INTEGER`, `GRB.BINARY`) and the direction of optimization (e.g., `GRB.MAXIMIZE`, `GRB.MINIMIZE`). Also, it is used to generate status constant indicating the result of optimization (e.g., `GRB.OPTIMAL`, `GRB.INFEASIBLE`).

In [2]:
import gurobipy as gp
from gurobipy import GRB

## Resource Assignment Problem
### Data
The list $R$ contains the names of the three resources: Carlos, Joe, and Monika.

The list $J$ contains the names of the job positions: Tester, Java Developer, and Architect.

$r \in R$: index and set of resources. The resource $r$ belongs to the set of resources $R$.

$j \in J$: index and set of jobs. The job $j$ belongs to the set of jobs $J$.

As can be easily seen in the "Problem Description", this problem is a typical assignement problem, which can be modeled as Minimum Cost Network Flow Problem (MCNFP, hereafter).
The following codes are for creating the nodes.

In [3]:
# Resource and job sets
R = ['Carlos', 'Joe', 'Monika']
J = ['Tester', 'JavaDeveloper', 'Architect']

The ability of each resource to perform each of the jobs is listed in the following matching scores table:

![scores](https://github.com/Gurobi/modeling-examples/blob/master/milp_tutorial/util/matching_score_data.PNG?raw=1)

For each resource $r$ and job $j$, there is a corresponding matching score $s$. The matching score $s$ can only take values between 0 and 100. That is, $s_{r,j} \in [0, 100]$ for all resources $r \in R$ and jobs $j \in J$. 

Note that each matching score, $s_{r,j}$, is basically "cost" in MCNFP's. Here we are trying to maximize the sum of scores. Therefore, in the context of MCNFP, we can think of scores as negative costs and we are trying to minimize the sum of scores.

We use the Gurobi Python ``multidict`` function to initialize one or more dictionaries with a single statement. The function takes a dictionary as its argument. The keys represent the possible combinations of resources and jobs.

In [4]:
# Matching score data
# negative cost in the context of MCNFP

combinations, scores = gp.multidict({
        ('Carlos', 'Tester'): 53,
        ('Carlos', 'JavaDeveloper'): 27,
        ('Carlos', 'Architect'): 13,
        ('Joe', 'Tester'): 80,
        ('Joe', 'JavaDeveloper'): 47,
        ('Joe', 'Architect'): 67,
        ('Monika', 'Tester'): 53,
        ('Monika', 'JavaDeveloper'): 73,
        ('Monika', 'Architect'): 47
    })

Here, "combinations" is a python list object, which stores the keys - 'decision variables.'

In [5]:
print("Combinations:", combinations)
print("Type of combinations:", type(combinations))

Combinations: <gurobi.tuplelist (9 tuples, 2 values each):
 ( Carlos , Tester        )
 ( Carlos , JavaDeveloper )
 ( Carlos , Architect     )
 ( Joe    , Tester        )
 ( Joe    , JavaDeveloper )
 ( Joe    , Architect     )
 ( Monika , Tester        )
 ( Monika , JavaDeveloper )
 ( Monika , Architect     )
>
Type of combinations: <class 'gurobipy.tuplelist'>


On the other hand, "scores" is a python dictionary object, which stores the values as a dictionary -  'cost coefficients.'

In [6]:
print("\nScores:", scores)
print("Type of scores:", type(scores))


Scores: {('Carlos', 'Tester'): 53, ('Carlos', 'JavaDeveloper'): 27, ('Carlos', 'Architect'): 13, ('Joe', 'Tester'): 80, ('Joe', 'JavaDeveloper'): 47, ('Joe', 'Architect'): 67, ('Monika', 'Tester'): 53, ('Monika', 'JavaDeveloper'): 73, ('Monika', 'Architect'): 47}
Type of scores: <class 'gurobipy.tupledict'>


Use the following constructor to create an empty model object, "m." 'RAP' is simply an argument to designate the name to the model, which stands for 'resource assignment problem.'

In [7]:
# Declare and initialize model
m = gp.Model('RAP')

Set parameter Username


## Decision variables

To solve this assignment problem, we need to identify which resource is assigned to which job. We introduce a decision variable for each possible assignment of resources to jobs. Therefore, we have 9 decision variables. \
(**Note.** each arc is a binary decision variable.)

To simplify the mathematical notation of the model formulation, we define the following indices for resources and jobs:

![variables](https://github.com/Gurobi/modeling-examples/blob/master/milp_tutorial/util/decision_variables.PNG?raw=1)

For example, $x_{2,1}$ is the decision variable associated with assigning the resource Joe to the job Tester. Therefore, decision variable $x_{r,j}$ equals 1 if resource $r \in R$  is assigned to job $j \in J$, and 0 otherwise.

The ``Model.addVars()`` method creates the decision variables for a ``Model`` object.
This method returns a Gurobi ``tupledict`` object that contains the newly created variables. We supply the ``combinations`` object as the first argument to specify the variable indices. The ``name`` keyword is used to specify a name for the newly created decision variables. By default, variables are assumed to be non-negative.

In [8]:
# Create decision variables for RAP model
x = m.addVars(combinations, name = 'assign')

## Job constraints

We now discuss the constraints associated with the jobs. These constraints need to ensure that each job is filled by exactly one resource.

The job constraint for the Tester position requires that resource 1 (Carlos), resource 2 (Joe), or resource 3 (Monika) is assigned to this job. This corresponds to the following constraint.

Constraint (Tester=1)

$$
x_{1,1} + x_{2,1} + x_{3,1} = 1
$$

Similarly, the constraints for the Java Developer and Architect positions can be defined as follows.

Constraint (Java Developer = 2)

$$
x_{1,2} + x_{2,2} + x_{3,2} = 1
$$

Constraint (Architect = 3)

$$
x_{1,3} + x_{2,3} + x_{3,3} = 1
$$

The job constraints are defined by the columns of the following table.

![jobs](https://github.com/Gurobi/modeling-examples/blob/master/milp_tutorial/util/jobs_constraints.PNG?raw=1)

In general, the constraint for the job Tester can defined as follows.

$$
x_{1,1} + x_{2,1} + x_{3,1} = \sum_{r=1}^{3 } x_{r,1} =  \sum_{r \in R} x_{r,1} = 1
$$

All of the job constraints can be defined in a similarly succinct manner. For each job $j \in J$, take the summation of the decision variables over all the resources. We can write the corresponding job constraint as follows.

$$
\sum_{r \in R} x_{r,j} = 1
$$

The ``Model.addConstrs()`` method of the Gurobi/Python API defines the job constraints of the ``Model`` object “m”. This method returns a Gurobi ``tupledict`` object that contains the job constraints. 
The first argument of this method, "x.sum(‘*’, j)", is the sum method and defines the LHS of the jobs constraints as follows:
For each job $j$ in the set of jobs $J$, take the summation of the decision variables over all the resources. The $==$  defines an equality constraint, and the number "1" is the RHS of the constraints.
These constraints are saying that exactly one resource should be assigned to each job.
The second argument is the name of this type of constraints.


In [9]:
# Create job constraints
jobs = m.addConstrs((x.sum('*', j) == 1 for j in J), name = 'job')

## Resource constraints

The constraints for the resources need to ensure that at most one job is assigned to each resource. That is, it is possible that not all the resources are assigned.

For example, we want a constraint that requires Carlos to be assigned to at most one of the jobs: either job 1 (Tester), job 2 (Java Developer ), or job 3 (Architect). We can write this constraint as follows.

Constraint (Carlos=1)

$$
x_{1, 1} + x_{1, 2} + x_{1, 3}  \leq 1.
$$

This constraint is less or equal than 1 to allow the possibility that Carlos is not assigned to any job. Similarly, the constraints for the resources Joe and Monika can be defined as follows:

Constraint (Joe=2) 

$$
x_{2, 1} + x_{2, 2} + x_{2, 3}  \leq 1.
$$

Constraint (Monika=3)

$$
x_{3, 1} + x_{3, 2} + x_{3, 3}  \leq 1.
$$

Observe that the resource constraints are defined by the rows of the following table.

![resources](https://github.com/Gurobi/modeling-examples/blob/master/milp_tutorial/util/resource_constraints.PNG?raw=1)

The constraint for the resource Carlos can be defined as follows.

$$
x_{1, 1} + x_{1, 2} + x_{1, 3} = \sum_{j=1}^{3 } x_{1,j} = \sum_{j \in J} x_{1,j} \leq 1.
$$

Again, each of these constraints can be written in a succinct manner. For each resource $r \in R$, take the summation of the decision variables over all the jobs. We can write the corresponding resource constraint as follows.

$$
\sum_{j \in J} x_{r,j} \leq  1.
$$

The ``Model.addConstrs()`` method of the Gurobi/Python API defines the resource constraints of the ``Model`` object “m”. 
The first argument of this method, "x.sum(r, ‘*’)", is the sum method and defines the LHS of the resource constraints as follows: For each resource $r$ in the set of resources $R$, take the summation of the decision variables over all the jobs.
The $<=$  defines a less or equal constraints, and the number “1” is the RHS of the constraints.
These constraints are saying that each resource can be assigned to at most 1 job.
The second argument is the name of this type of constraints.


In [10]:
# Create resource constraints
resources = m.addConstrs((x.sum(r, '*') <= 1 for r in R), name = 'resource')

## Objective function

The objective function is to maximize the total matching score of the assignments that satisfy the job and resource constraints. 

For the Tester job, the matching score is $53x_{1,1}$, if resource Carlos is assigned, or $80x_{2,1}$, if resource Joe is assigned, or $53x_{3,1}$, if resource Monika is assigned.
Consequently, the matching score for the Tester job is as follows, where only one term in this summation will be nonzero.

$$
53x_{1,1} + 80x_{2,1} + 53x_{3,1}. 
$$

Similarly, the matching scores for the Java Developer and Architect jobs are defined as follows. The matching score for the Java Developer job is:

$$
27x_{1, 2} + 47x_{2, 2} + 73x_{3, 2}.
$$

The matching score for the Architect job is:

$$
13x_{1, 3} + 67x_{2, 3} + 47x_{3, 3}.
$$

The total matching score is the summation of each cell in the following table.

![objfcn](https://github.com/Gurobi/modeling-examples/blob/master/milp_tutorial/util/objective_function.PNG?raw=1)

The goal is to  maximize the total matching score of the assignments. Therefore, the objective function is defined as follows.

\begin{equation}
\text{Maximize} \quad (53x_{1,1} + 80x_{2,1} + 53x_{3,1}) \; +
\end{equation}

\begin{equation}
\quad (27x_{1, 2} + 47x_{2, 2} + 73x_{3, 2}) \; +
\end{equation}

\begin{equation}
\quad (13x_{1, 3} + 67x_{2, 3} + 47x_{3, 3}).
\end{equation}

Each term in parenthesis in the objective function can be expressed as follows.

\begin{equation}
(53x_{1,1} + 80x_{2,1} + 53x_{3,1}) = \sum_{r \in R} s_{r,1}x_{r,1}.
\end{equation}

\begin{equation}
(27x_{1, 2} + 47x_{2, 2} + 73x_{3, 2}) = \sum_{r \in R} s_{r,2}x_{r,2}.
\end{equation}

\begin{equation}
(13x_{1, 3} + 67x_{2, 3} + 47x_{3, 3}) = \sum_{r \in R} s_{r,3}x_{r,3}.
\end{equation}

Hence, the objective function can be concisely written as:

\begin{equation}
\text{Maximize} \quad \sum_{j \in J} \sum_{r \in R} s_{r,j}x_{r,j}.
\end{equation}

The ``Model.setObjective()`` method of the Gurobi/Python API defines the objective function of the ``Model`` object “m”. The objective expression is specified in the first argument of this method.
Notice that both the matching score parameters “score” and the assignment decision variables “x” are defined over the “combinations” keys. Therefore, we use the method “x.prod(score)” to obtain the summation of the elementwise multiplication of the "score" matrix and the "x" variable matrix.
The second argument, ``GRB.MAXIMIZE``, is the optimization "sense." In this case, we want to *maximize* the total matching scores of all assignments.

**Note.** maximizing scores is equivalent to minimizing the sum of costs if we take scores as negative costs.\
**Note.** "scores" object is a dictionary with keys of decision variables and values of costs.

In [11]:
# Objective: maximize total matching score of all assignments
m.setObjective(x.prod(scores), GRB.MAXIMIZE)

Use the `write()` method of Gurobi/Python API to write the model formulation to a file names "RAP.lp."

In [12]:
# Save model for insepction
m.write('RAP.lp')

We can check the model specification using below code.

In [13]:
# open the file in reade mode
# store the lp file as a file object
# with statement ensures the file is propley closed after executing the code block

with open('RAP.lp', 'r') as file:
    content = file.read()
    print(content)

\ Model RAP
\ LP format - for model browsing. Use MPS format to capture full model detail.
Maximize
  53 assign[Carlos,Tester] + 27 assign[Carlos,JavaDeveloper]
   + 13 assign[Carlos,Architect] + 80 assign[Joe,Tester]
   + 47 assign[Joe,JavaDeveloper] + 67 assign[Joe,Architect]
   + 53 assign[Monika,Tester] + 73 assign[Monika,JavaDeveloper]
   + 47 assign[Monika,Architect]
Subject To
 job[Tester]: assign[Carlos,Tester] + assign[Joe,Tester]
   + assign[Monika,Tester] = 1
 job[JavaDeveloper]: assign[Carlos,JavaDeveloper]
   + assign[Joe,JavaDeveloper] + assign[Monika,JavaDeveloper] = 1
 job[Architect]: assign[Carlos,Architect] + assign[Joe,Architect]
   + assign[Monika,Architect] = 1
 resource[Carlos]: assign[Carlos,Tester] + assign[Carlos,JavaDeveloper]
   + assign[Carlos,Architect] <= 1
 resource[Joe]: assign[Joe,Tester] + assign[Joe,JavaDeveloper]
   + assign[Joe,Architect] <= 1
 resource[Monika]: assign[Monika,Tester] + assign[Monika,JavaDeveloper]
   + assign[Monika,Architect] <= 1


**Note.**\
One might think that additional constraint limiting decision variables to be binary is missing. But such constraint is redundant thereby this problem can be relaxed into Linear Programming. This model satisfies integrality property so the optimial solutions are integers. Also, due to the specific constraints forcing one-to-one matches, the optimal solutions are binary.

Use "optimize()" method to solve the problem.

In [14]:
# Run optimization engine
m.optimize()

Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (win64 - Windows 10.0 (19045.2))

CPU model: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 6 physical cores, 12 logical processors, using up to 12 threads

Optimize a model with 6 rows, 9 columns and 18 nonzeros
Model fingerprint: 0xb343b6eb
Coefficient statistics:
  Matrix range     [1e+00, 1e+00]
  Objective range  [1e+01, 8e+01]
  Bounds range     [0e+00, 0e+00]
  RHS range        [1e+00, 1e+00]
Presolve time: 0.01s
Presolved: 6 rows, 9 columns, 18 nonzeros

Iteration    Objective       Primal Inf.    Dual Inf.      Time
       0    4.6000000e+32   1.800000e+31   4.600000e+02      0s
       5    1.9300000e+02   0.000000e+00   0.000000e+00      0s

Solved in 5 iterations and 0.01 seconds (0.00 work units)
Optimal objective  1.930000000e+02


`Model.getVars()` method retrieves a list of all variables in the model. The ``.x`` variable attribute is used to query solution values and the ``.varName`` attribute is used to query the name of the decision variables.  

In [15]:
# Display optimal values of decision variables
for v in m.getVars():
    if v.x > 1e-6:
        print(v.varName, v.x)

# Display optimal total matching score
print('Total matching score: ', m.objVal)

assign[Carlos,Tester] 1.0
assign[Joe,Architect] 1.0
assign[Monika,JavaDeveloper] 1.0
Total matching score:  193.0


The optimal assignment is to assign:

* Carlos to the Tester job, with a matching score of 53
* Joe to the Architect job, with a matching score of 67
* Monika to the Java Developer job, with a matching score of 73.

The maximum total matching score is 193.

### Summary for RAP



\begin{equation}
\text{Maximize} \quad \sum_{j \in J} \sum_{r \in R} s_{r,j}x_{r,j} = \quad (53x_{1,1} + 80x_{2,1} + 53x_{3,1}) \; +
\quad (27x_{1, 2} + 47x_{2, 2} + 73x_{3, 2}) \; +
\quad (13x_{1, 3} + 67x_{2, 3} + 47x_{3, 3})
\end{equation}

\begin{align*}
\text{s.t.} &\sum_{r \in R} x_{r,j} = 1, \quad \forall j \in J \quad \text{(job constraints)}\\
            &\sum_{j \in J} x_{r,j} \leq  1, \quad \forall r \in R \quad \text{(resource constraints)}
\end{align*}



```python
# Resource and job sets
R = ['Carlos', 'Joe', 'Monika']
J = ['Tester', 'JavaDeveloper', 'Architect']

# Matching score data
combinations, scores = gp.multidict({
        ('Carlos', 'Tester'): 53,
        ('Carlos', 'JavaDeveloper'): 27,
        ('Carlos', 'Architect'): 13,
        ('Joe', 'Tester'): 80,
        ('Joe', 'JavaDeveloper'): 47,
        ('Joe', 'Architect'): 67,
        ('Monika', 'Tester'): 53,
        ('Monika', 'JavaDeveloper'): 73,
        ('Monika', 'Architect'): 47
    })

# Declare and initialize model
m = gp.Model('RAP')

# Create decision variables for RAP model
x = m.addVars(combinations, name = 'assign') #integrality property - no need to explicitly add binary condition

# Create job constraints
jobs = m.addConstrs((x.sum('*', j) == 1 for j in J), name = 'job')

# Create resource constraints
resources = m.addConstrs((x.sum(r, '*') <= 1 for r in R), name = 'resource')

# Objective: maximize total matching score of all assignments
m.setObjective(x.prod(scores), GRB.MAXIMIZE)

# Save model for insepction
m.write('RAP.lp')

# inspection
with open('RAP.lp', 'r') as file:
    content = file.read()
    print(content)

# Run optimization engine
m.optimize()

# Display optimal values of decision variables
for v in m.getVars():
    if v.x > 1e-6:
        print(v.varName, v.x)

# Display optimal total matching score
print('Total matching score: ', m.objVal)

## Resource Assignment Problem with a budget constraint

Now, assume there is a fixed cost $C_{r,j}$ associated with assigning a resource $r \in R$ to job $j \in J$. Assume also that there is a limited budget $B$ that can be used for job assignments.

The cost of assigning Carlos, Joe, or Monika to any of the jobs is $\$1,000$ , $\$2,000$ , and $\$3,000$  respectively. The available budget is $\$5,000$.

### Data

The list $R$ contains the names of the three resources: Carlos, Joe, and Monika.
The list $J$ contains the names of the job positions: Tester, Java Developer, and Architect.

The Gurobi Python ``multidict`` function initialize two dictionaries: 
* "scores" defines the matching scores for each resource and job combination.
* "costs" defines the fixed cost associated of assigning a resource to a job.

In [16]:
# Resoucre and job sets
# nodes in Assignment problems
R = ['Carlos', 'Joe', 'Monika']
J = ['Tester', 'JavaDeveloper', 'Architect']

# Matching score data
# Cost is given in thousands of dollars
combinations, scores, costs = gp.multidict({
        ('Carlos', 'Tester'): [53, 1],
        ('Carlos', 'JavaDeveloper'): [27, 1],
        ('Carlos', 'Architect'): [13,1],
        ('Joe', 'Tester'): [80, 2],
        ('Joe', 'JavaDeveloper'): [47, 2],
        ('Joe', 'Architect'): [67, 2],
        ('Monika', 'Tester'): [53, 3] ,
        ('Monika', 'JavaDeveloper'): [73, 3],
        ('Monika', 'Architect'): [47, 3]
    })

# Available budget (thousands of dollars)
budget = 5

The following constructor creates an empty Model object “m”. The Model object “m” holds a single optimization problem. It consists of a set of variables, a set of constraints, and the objective function.

In [17]:
# Declare and initialize model
m = gp.Model('RAP2')

### Decision variables

The decision variable $x_{r,j}$ is 1 if $r \in R$ is assigned to job $j \in J$, and 0 otherwise.

The ``Model.addVars()`` method defines the decision variables for the model object “m”.  

Because there is a budget constraint, it is possible that not all of the jobs will be filled. To account for this, we define a new decision variable that indicates whether or not a job is filled.

Let $g_{j}$ be equal 1 if job $j \in J$ is not filled, and 0 otherwise. This variable is a gap variable that indicates that a job cannot be filled.

***Remark:*** For the previous formulation of the RAP, we defined the assignment variables as non-negative and continuous which is the default value of the ``vtype`` argument of the ``Model.addVars()`` method.
However, in this extension of the RAP, because of the budget constraint we added to the model, we need to explicitly define these variables as binary. The ``vtype=GRB.BINARY`` argument of the ``Model.addVars()`` method defines the assignment variables as binary.

In [18]:
# Create decision variables for the RAP model
x = m.addVars(combinations, vtype = GRB.BINARY, name = 'assign')

# Create gap variables for the RAP model
g = m.addVars(J, name = 'gap')

### Job constraints

Since we have a limited budget to assign resources to jobs, it is possible that not all the jobs can be filled. For the job constraints, there are two possibilities either a resource is assigned to fill the job, or this job cannot be filled and we need to declare a gap. This latter possibility is captured by the decision variable $g_j$. Therefore, the job constraints are written as follows. 

For each job $j \in J$, exactly one resource must be assigned to the job, or the corresponding $g_j$ variable must be set to 1:

$$
\sum_{r \: \in \: R} x_{r,\; j} + g_{j} = 1.
$$


In [19]:
# Create job constraints
jobs = m.addConstrs((x.sum('*',j) + g[j]  == 1 for j in J), name='job')

### Resource constraints

The constraints for the resources need to ensure that at most one job is assigned to each resource. That is, it is possible that not all the resources are assigned. Therefore, the resource constraints are written as follows.

For each resource $r \in R$, at most one job can be assigned to the resource:

$$
\sum_{j \: \in \: J} x_{r,\; j} \leq 1.
$$

In [20]:
# Create resource constraints
resources = m.addConstrs((x.sum(r,'*') <= 1 for r in R), name='resource')

### Budget constraint

This constraint ensures that the cost of assigning resources to fill job requirements do not exceed the budget available. The costs of assignment and budget are in thousands of dollars.

The cost of filling the Tester job is $1x_{1,1}$, if resource Carlos is assigned, or $2x_{2,1}$, if resource Joe is assigned, or $3x_{3,1}$, if resource Monika is assigned.
Consequently, the cost of filling the Tester job is as follows, where at most one term in this summation will be nonzero.

$$
1x_{1,1} + 2x_{2,1} + 3x_{3,1}. 
$$

Similarly, the cost of filling the Java Developer and Architect jobs are defined as follows. The cost of filling the Java Developer job is:

$$
1x_{1, 2} + 2x_{2, 2} + 3x_{3, 2}.
$$

The cost of filling the Architect job is:

$$
1x_{1, 3} + 2x_{2, 3} + 3x_{3, 3}.
$$

Hence, the total cost of filling the jobs should be less or equal than the budget available.

\begin{equation}
(1x_{1,1} + 2x_{2,1} + 3x_{3,1}) \; +
\end{equation}

\begin{equation}
(1x_{1, 2} + 2x_{2, 2} + 3x_{3, 2}) \; +
\end{equation}

\begin{equation}
(1x_{1, 3} + 2x_{2, 3} + 3x_{3, 3}) \leq 5
\end{equation}

Each term in parenthesis in the budget constraint can be expressed as follows.

\begin{equation}
(1x_{1,1} + 2x_{2,1} + 3x_{3,1}) = \sum_{r \in R} C_{r,1}x_{r,1}.
\end{equation}

\begin{equation}
(1x_{1, 2} + 2x_{2, 2} + 3x_{3, 2}) = \sum_{r \in R} C_{r,2}x_{r,2}.
\end{equation}

\begin{equation}
(1x_{1, 3} + 2x_{2, 3} + 3x_{3, 3}) = \sum_{r \in R} C_{r,3}x_{r,3}.
\end{equation}

Therefore, the budget constraint can be concisely written as:

\begin{equation}
\sum_{j \in J} \sum_{r \in R} C_{r,j}x_{r,j} \leq B.
\end{equation}

The ``Model.addConstr()`` method of the Gurobi/Python API defines the budget constraint of the ``Model`` object “m”. 
The first argument of this method, "x.prod(costs)", is the prod method and defines the LHS of the budget constraint. The $<=$ defines a less or equal constraint, and the budget amount available is the RHS of the constraint.
This constraint is saying that the total cost of assigning resources to fill jobs requirements cannot exceed the budget available.
The second argument is the name of this constraint.

In [21]:
budget = m.addConstr((x.prod(costs) <= budget), name='budget')

## Objective function

The objective function is similar to the RAP. The first term in the objective is the total matching score of the assignments. In this extension of the RAP, it is possible that not all jobs are filled; however, we want to heavily penalize this possibility. For this purpose, we have a second term in the objective function that takes the summation of the gap variables over all the jobs and multiply it by a big penalty $M$.

Observe that the maximum value of a matching score is 100, and the value that we give to $M$ is 101. The rationale behind the value of $M$ is that having gaps heavily deteriorates the total matching scores value.

Consequently, the objective function is to maximize the total matching score of the assignments minus the penalty associated of having gap variables with a value equal to 1.

$$
\max \; \sum_{j \; \in \; J} \sum_{r \; \in \; R} s_{r,j}x_{r,j} -M \sum_{j \in J} g_{j}
$$

In [22]:
# Penalty for not filling a job position
M = 101

In [23]:
# Objective: maximize total matching score of assignments
# Unfilled jobs are heavily penalized
m.setObjective(x.prod(scores) - M*g.sum(), GRB.MAXIMIZE)

In [24]:
# Save model for insepction
m.write('RAP2.lp')

# inspection
with open('RAP2.lp', 'r') as file:
    content = file.read()
    print(content)

\ Model RAP2
\ LP format - for model browsing. Use MPS format to capture full model detail.
Maximize
  53 assign[Carlos,Tester] + 27 assign[Carlos,JavaDeveloper]
   + 13 assign[Carlos,Architect] + 80 assign[Joe,Tester]
   + 47 assign[Joe,JavaDeveloper] + 67 assign[Joe,Architect]
   + 53 assign[Monika,Tester] + 73 assign[Monika,JavaDeveloper]
   + 47 assign[Monika,Architect] - 101 gap[Tester] - 101 gap[JavaDeveloper]
   - 101 gap[Architect]
Subject To
 job[Tester]: assign[Carlos,Tester] + assign[Joe,Tester]
   + assign[Monika,Tester] + gap[Tester] = 1
 job[JavaDeveloper]: assign[Carlos,JavaDeveloper]
   + assign[Joe,JavaDeveloper] + assign[Monika,JavaDeveloper]
   + gap[JavaDeveloper] = 1
 job[Architect]: assign[Carlos,Architect] + assign[Joe,Architect]
   + assign[Monika,Architect] + gap[Architect] = 1
 resource[Carlos]: assign[Carlos,Tester] + assign[Carlos,JavaDeveloper]
   + assign[Carlos,Architect] <= 1
 resource[Joe]: assign[Joe,Tester] + assign[Joe,JavaDeveloper]
   + assign[Joe,

In [25]:
# Run optimization engine
m.optimize()

Gurobi Optimizer version 11.0.3 build v11.0.3rc0 (win64 - Windows 10.0 (19045.2))

CPU model: Intel(R) Core(TM) i7-10750H CPU @ 2.60GHz, instruction set [SSE2|AVX|AVX2]
Thread count: 6 physical cores, 12 logical processors, using up to 12 threads

Optimize a model with 7 rows, 12 columns and 30 nonzeros
Model fingerprint: 0xa1231a12
Variable types: 3 continuous, 9 integer (9 binary)
Coefficient statistics:
  Matrix range     [1e+00, 3e+00]
  Objective range  [1e+01, 1e+02]
  Bounds range     [1e+00, 1e+00]
  RHS range        [1e+00, 5e+00]
Presolve time: 0.00s
Presolved: 7 rows, 12 columns, 30 nonzeros
Variable types: 0 continuous, 12 integer (12 binary)
Found heuristic solution: objective 52.0000000

Root relaxation: objective 1.350000e+02, 4 iterations, 0.00 seconds (0.00 work units)

    Nodes    |    Current Node    |     Objective Bounds      |     Work
 Expl Unexpl |  Obj  Depth IntInf | Incumbent    BestBd   Gap | It/Node Time

     0     0  135.00000    0    2   52.00000  135.0

The definition of the objective function includes the penalty of no filling jobs. However, we are interested in the optimal total matching score value when not all the jobs are filled. For this purpose, we need to compute the total matching score value using the matching score values $s_{r,j}$ and the assignment decision variables $x_{r,j}$.

In [26]:
# Compute total matching score from assignment variables
total_matching_score = 0
for r, j in combinations:
    if x[r, j].x > 1e-6:
        print(x[r, j].varName, x[r, j].x) 
        total_matching_score += scores[r, j]*x[r, j].x

print('Total matching score: ', total_matching_score)  

assign[Joe,Tester] 1.0
assign[Monika,JavaDeveloper] 1.0
Total matching score:  153.0


In [27]:
# Breakdown of results
matching_score = sum(scores[r, j] * x[r, j].x for r, j in combinations if x[r, j].x > 1e-6)
penalty = M * sum(g[j].x for j in J)

print(f"Total matching score (assignments): {matching_score}")
print(f"Total penalty (unfilled jobs): {penalty}")
print(f"Final objective value: {matching_score - penalty}")

Total matching score (assignments): 153.0
Total penalty (unfilled jobs): 101.0
Final objective value: 52.0


### Analysis

Recall that the budget is $\$5,000$, and the total  cost associated of allocating the three resources is $\$6,000$. This means that there is not enough budget to allocate the three resources we have. Consequently, the Gurobi Optimizer must choose two resources to fill the jobs demand, leave one job unfilled, and maximize the total matching scores. Notice that the two top matching scores are 80% (Joe for the Tester job) and 73% (Monika for the Java Developer job). Also, notice that the lowest score is 13% (Carlos for the Architect job). Assigning Joe to the Tester job, Monika to the Java Developer job, and nobody to the Architect job costs $\$5,000$  and yields a total matching score of 153. This is the optimal solution found by the Gurobi Optimizer.

### Cleaning up the memory

`m.dispose()` releases all memory and resources assigned to `m` to prevent memory leaks.
`gp.disposeDefaultEnv()` is for a global cleanup. It is used when you are completely done with using Gurobi.

In [28]:
m.dispose()
gp.disposeDefaultEnv()

Freeing default Gurobi environment


### Summary for RAP2 (Budget Constraint)



\begin{equation}
\text{Maximize} \quad \sum_{j \; \in \; J} \sum_{r \; \in \; R} s_{r,j}x_{r,j} -M \sum_{j \in J} g_{j}
\end{equation}

\begin{align*}
\text{s.t.} \quad &\sum_{r \: \in \: R} x_{r,\; j} + g_{j} = 1, \quad \forall j \in J     &\text{(job constraints)}\\
                  &\sum_{j \in J} x_{r,j} \leq  1, \quad \forall r \in R & \text{(resource constraints)}\\
                  &\sum_{j \in J} \sum_{r \in R} c_{r,j}x_{r,j} \leq 5   & \text{(budget constraint)} \\ 
                  &\text{where} \quad c_{1,j} = 1, \; c_{2,j} = 2, \; c_{3,j} = 3, \quad \forall j \in J \\
                  &\text{and} \quad x_{r, j}, g_{r, j} \; \text{: binary} \quad ,\forall r \in R, \; j \in J
\end{align*}



```python
# Resource and job sets
R = ['Carlos', 'Joe', 'Monika']
J = ['Tester', 'JavaDeveloper', 'Architect']

# Matching score data
# Cost is given in thousands of dollars
combinations, scores, costs = gp.multidict({
        ('Carlos', 'Tester'): [53, 1],
        ('Carlos', 'JavaDeveloper'): [27, 1],
        ('Carlos', 'Architect'): [13,1],
        ('Joe', 'Tester'): [80, 2],
        ('Joe', 'JavaDeveloper'): [47, 2],
        ('Joe', 'Architect'): [67, 2],
        ('Monika', 'Tester'): [53, 3] ,
        ('Monika', 'JavaDeveloper'): [73, 3],
        ('Monika', 'Architect'): [47, 3]
    })

# Available budget (thousands of dollars)
budget = 5

# Declare and initialize model
m = gp.Model('RAP2')

# Create decision variables for the RAP model
x = m.addVars(combinations, vtype = GRB.BINARY, name = 'assign')

# Create gap variables for the RAP model
# gap = 1 when the job is unfulfilled
g = m.addVars(J, name = 'gap')

# Create job constraints
jobs = m.addConstrs((x.sum('*',j) + g[j]  == 1 for j in J), name='job')

# Create resource constraints
resources = m.addConstrs((x.sum(r,'*') <= 1 for r in R), name='resource')

# Create budget constraint
budget = m.addConstr((x.prod(costs) <= budget), name='budget')

# Penalty for not filling a job position
M = 101

# Objective: maximize total matching score of assignments
# Unfilled jobs are heavily penalized
m.setObjective(x.prod(scores) - M*g.sum(), GRB.MAXIMIZE)

# Save model for insepction
m.write('RAP2.lp')

# inspection
with open('RAP2.lp', 'r') as file:
    content = file.read()
    print(content)

# Run optimization engine
m.optimize()

# Compute total matching score from assignment variables
total_matching_score = 0
for r, j in combinations:
    if x[r, j].x > 1e-6:
        print(x[r, j].varName, x[r, j].x) 
        total_matching_score += scores[r, j]*x[r, j].x

print('Total matching score: ', total_matching_score)  

# Breakdown of results
matching_score = sum(scores[r, j] * x[r, j].x for r, j in combinations if x[r, j].x > 1e-6)
penalty = M * sum(g[j].x for j in J)

print(f"Total matching score (assignments): {matching_score}")
print(f"Total penalty (unfilled jobs): {penalty}")
print(f"Final objective value: {matching_score - penalty}")